sdk/nodejs/remote/index.ts

```diff
@@ -0,0 +1,107 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import * as cp from "child_process";
+import * as grpc from "grpc";
+import * as pulumi from "../";
+import * as runtime from "../runtime";
+
+//tslint:disable
+const gstruct = require("google-protobuf/google/protobuf/struct_pb.js");
+const runtimeServiceProto = require("../proto/runtime_grpc_pb.js");
+const runtimeProto = require("../proto/runtime_pb.js");
+
+/**
+ * ProxyComponentResource is the abstract base class for proxies around component resources.
+ *
+ * TODO: This should move into the core NodeJS SDK.
+ */
+export abstract class ProxyComponentResource extends pulumi.ComponentResource {
+    constructor(
+        t: string,
+        name: string,
+        libraryPath: string,
+        libraryName: string,
+        inputs: pulumi.Inputs,
+        outputs: Record<string, undefined>,
```

Comment:
cute.

Reply:
outputs is effectively a set/list of names.  maybe just `outputNames: string[]`?

Reply:
I actually preferred the consistency of this for user code between inputs and outputs - while still getting type checking.  Also makes the implementation here very simple/obvious/aligned-with-pulumi-resource-model.